### DISCUSSION TITLE
How to k3s automatic update bootstrap data in the external database when user manual changes the agent token with k3s server CLI for start the K3S Cluster.

### DISCUSSION THREAD

**Discussion by laiminhtrung1997** [k3s-io/k3s#8180] [General]
**Environmental Info:**
K3s Version: v1.27.2+k3s1

Cluster Configuration: 2 servers

**Describe the bug:**
I start the K3S clusters with the external database using Postgresql. I used the same server token as many times startup the K3S Cluster but the agent token does not. I manually change the agent token each time I start the K3S Cluster.
As far as I understand the source code when the K3S Cluster starts for the first time, the directory /var/lib/rancher/k3s/server/cred does not exist including the passwd file containing the server token and the agent token, then the k3s will generate the passwd file with the value of K3S_TOKEN and K3S_AGENT_TOKEN environment, then write down to the bootstrap data in the database.
Because I never change the server token but the agent token has changed each time I restart the K3S Cluster, the bootstrap key /bootstrap/xxxxxxxxxxxx will be never changed. Therefore, the k3s will not write down the newest agent token to the bootstrap data in the database as I expected.
So, I have 2 questions
1./ What is the purpose of the k3s to store the bootstrap data such as certs, keys, etc both on disk and database?
2./ With my scenario, What can I do to make the k3s can write down the newest agent token to the database? Is it a bug?

**Steps To Reproduce:**
1./ Start the K3S Cluster for the first time with the environments:
K3S_TOKEN: 123
K3S_AGENT_TOKEN: 456

2./ Stop then restart the K3S Cluster for the second time with environments:
K3S_TOKEN: 123
K3S_AGENT_TOKEN: 789

**Expected behavior:**
The K3S Cluster start successfully.
The bootstrap data in the database contains the new agent token which is 789.

**Actual behavior:**
The K3S Cluster cannot start.
The bootstrap data in the database contains the old agent token which is 456.

**Additional context/logs:**
```
time="2023-07-18T04:43:40Z" level=warning msg="Webhooks and apiserver aggregation may not function properly without an agent; please set egress-selector-mode to 'cluster' or 'pod'"
time="2023-07-18T04:43:40Z" level=info msg="Starting k3s v1.27.2+k3s1"
time="2023-07-18T04:43:40Z" level=info msg="Configuring pgx database connection pooling: maxIdleConns=2, maxOpenConns=0, connMaxLifetime=0s"
time="2023-07-18T04:43:40Z" level=info msg="Configuring database table schema and indexes, this may take a moment..."
time="2023-07-18T04:43:40Z" level=info msg="Database tables and indexes are up to date"
time="2023-07-18T04:43:41Z" level=info msg="Kine available at unix://kine.sock"
time="2023-07-18T04:43:41Z" level=info msg="Reconciling bootstrap data between datastore and disk"
time="2023-07-18T04:43:41Z" level=fatal msg="/var/lib/rancher/k3s/server/cred/passwd newer than datastore and could cause a cluster outage. Remove the file(s) from disk and restart to be recreated from datastore."
```

**Comment by brandond** [comment]
Token rotation (changing the static server or agent tokens) is not currently supported. There is an open feature request tracking the addition of support for this feature.

If you want to use ephemeral bootstrap tokens, see  https://docs.k3s.io/cli/token#bootstrap

**Comment by laiminhtrung1997** [comment]
Dear @brandond
Thank you for your support. I am so waiting for this new feature.
Best regards.

**Comment by laiminhtrung1997** [comment]
Dear @brandond
I tried to use the ephemeral bootstrap tokens like you said but is it to use for joining to existing node?
I want to start the K3S cluster successfully when I change the K3S_AGENT_TOKEN each time I run the command k3s server. Are there any things else to do this?

**Comment by brandond** [comment]
> I tried to use the ephemeral bootstrap tokens like you said but is it to use for joining to existing node?

I don't think I understand the question. If the node is already part of the cluster, then it doesn't need to join.

> I want to start the K3S cluster successfully when I change the K3S_AGENT_TOKEN each time I run the command k3s server.

Don't do that; as I mentioned above changing the static tokens is not currently supported. If you want to change or invalidate the join tokens, use ephemeral bootstrap tokens.

**Comment by caroline-suse-rancher** [comment]
I'm converting this to a discussion since we're already tracking this feature request

**Reply by laiminhtrung1997** [reply]
Dear @caroline-suse-rancher 
Can you tell me when this feature request can be released?

**Comment by laiminhtrung1997** [comment]
Dear @caroline-suse-rancher
Any news about this change request to solve this issue?

**Reply by brandond** [reply]
This has been covered in the docs for a few months: https://docs.k3s.io/cli/token#k3s-token-rotate
